### PR TITLE
Support full-line comments in fparser2

### PR DESCRIPTION
**Pull request by arporter**
This PR is for work done under #68.

### CODE REVIEW

**Comment by arporter** [comment]
I propose to leave the bug when generating fixed-format comments since that's not new and this PR is big enough already. Ready for first review by @rupertford.

File: src/fparser/api.py [hunk 1]
```diff
@@ -224,7 +226,8 @@ def parse(input, isfree=None, isstrict=None, include_dirs=None,
     get_reader
     """
     from .parsefortran import FortranParser
-    reader = get_reader(input, isfree, isstrict, include_dirs, source_only)
+    reader = get_reader(input, isfree, isstrict, include_dirs, source_only,
+                        ignore_comments=ignore_comments)
```
Comment (rupertford) [inline src/fparser/api.py]: Do we want to default to ignoring comments, now that we support them?

File: src/fparser/Fortran2003.py [hunk 2]
```diff
@@ -156,20 +158,23 @@ class Base(ComparableMixin):
...
 
-        :param cls: the class of object to create
-        :type cls: :py:type:`type`
+        :param type cls: the class of object to create
         :param string: (source of) Fortran string to parse
         :type string: str or :py:class:`FortranReaderBase`
         :param parent_cls: the parent class of this object
         :type parent_cls: :py:type:`type`
         """
+        from . import readfortran
```
Comment (rupertford) [inline src/fparser/Fortran2003.py]: I thought relative imports were bad?

File: src/fparser/Fortran2003.py [hunk 3]
```diff
@@ -268,38 +287,87 @@ class BlockBase(Base):
...
+        Checks whether the content in reader matches the given
+        type of block statement (e.g. DO..END DO, IF...END IF etc.)
+
+        :param type startcls: The class marking the beginning of the block
+        :param list subclasses: List of classes that can be children of
+                                the block
+        :param type endcls: The class marking the end of the block
+        :param reader: Content to check for match
+        :type reader: str or instance of :py:class:`FortranReaderBase`
+        :param bool match_labels: TBD
```
Comment (rupertford) [inline src/fparser/Fortran2003.py]: Lots of TBD entries here. Please either justify why or update.

File: src/fparser/Fortran2003.py [hunk 4]
```diff
@@ -268,38 +287,87 @@ class BlockBase(Base):
...
+        :param bool enable_case_construct_hook: TBD
+
+        :return: instance of startcls or None if no match is found
+        :rtype: startcls
+        '''
         assert isinstance(reader, FortranReaderBase), repr(reader)
         content = []
+
         if startcls is not None:
+            # Deal with any preceding comments
```
Comment (rupertford) [inline src/fparser/Fortran2003.py]: What about comments after the block? I can see that these might be picked up by the next block but what about the last block in a program. Is it guaranteed that we never get into a case where comments are not dealt with?

File: src/fparser/api.py [hunk 5]
```diff
@@ -135,10 +135,12 @@ def get_reader(input, isfree=None, isstrict=None, include_dirs=None,
                 isstrict = True
             return parse(c_input, isfree, isstrict, include_dirs)
         reader = FortranFileReader(input, include_dirs=include_dirs,
-                                   source_only=source_only)
+                                   source_only=source_only,
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/fparser/api.py]: Good point. Done.

File: src/fparser/api.py (hunk 1 above)
Comment (arporter) [reply src/fparser/api.py]: I think so given that that was effectively the behaviour before.

File: src/fparser/readfortran.py [hunk 6]
```diff
@@ -217,7 +221,9 @@ class Line(object):
 
     def __init__(self, line, linenospan, label, name, reader):
         self.line = line.strip()
-        assert self.line, repr((line, linenospan, label))
+        if not self.line:
+            raise Exception("Got empty line: '{0}'. linenospan={1}, "
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/fparser/readfortran.py]: Good point. Done.

File: src/fparser/readfortran.py [hunk 7]
```diff
@@ -578,16 +605,16 @@ def get_single_line(self, ignore_empty=False, ignore_comments=False):
         # expand tabs, replace special symbols, get rid of nl characters
         line = line.expandtabs().replace('\xa0',' ').rstrip()
         self.source_lines.append(line)
-
-        if ignore_comments and _is_fix_comment(line, isstrict=self.isstrict):
-            return self.get_single_line(ignore_empty, ignore_comments)
+        if ignore_comments and (self.isfix or self.isf77):
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/fparser/readfortran.py]: I had to add the isfix and isf77 checks in order that I can safely call the _is_fix_comment() routine.
I've added a comment to explain this if case.

File: src/fparser/readfortran.py [hunk 8]
```diff
@@ -666,41 +702,47 @@ def next(self, ignore_comments = False):
...
-                return self.reader.next(ignore_comments = ignore_comments)
+                self.reader = FortranFileReader(path,
+                                                include_dirs=include_dirs,
+                                                ignore_comments=ignore_comments)
+                return self.reader.next(ignore_comments=ignore_comments)
             return item
         except StopIteration:
             raise
-        except:
+        except Exception as err:
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/fparser/readfortran.py]: I agree. In fact it has already caused me significant grief in debugging the issues I had after merging with trunk. However, I don't know what specific things we should be catching at this point and this feels like a separate issue.

File: src/fparser/readfortran.py [hunk 9]
```diff
@@ -666,41 +702,47 @@ def next(self, ignore_comments = False):
...
             message = self.format_message('FATAL ERROR',
                                           'while processing line',
                                           self.linecount, self.linecount)
             logging.getLogger(__name__).critical(message)
             message = ''.join(('Traceback\n',
                               ''.join( traceback.format_stack())))
             logging.getLogger(__name__).debug(message)
+            logging.getLogger(__name__).debug(str(err))
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/fparser/readfortran.py]: No, an enhancement - you now get the text of the original exception logged.

File: src/fparser/readfortran.py [hunk 10]
```diff
@@ -881,44 +897,43 @@ def handle_cf2py_start(self, line):
...
-        -------
-        line_with_no_comments : str
-        quotechar : {None, str}
-        had_comment : bool
-
-        Notes
-        -----
-        In-line comments are separated from line and put back to fifo
-        sequence where it will be processed as comment line.
+        Any in-line comment is extracted from line. If
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/fparser/readfortran.py]: Done.

File: src/fparser/readfortran.py [hunk 11]
```diff
@@ -944,6 +959,8 @@ def handle_inline_comment(self, line, lineno, quotechar=None):
         return ''.join(noncomment_items), newquotechar, had_comment
 
     def handle_multilines(self, line, startlineno, mlstr):
+        '''
+        '''
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/fparser/readfortran.py]: Oops :-) I think I started doing that when I thought I was going to have to modify a method but it seems that in this case I never actually did...

File: src/fparser/readfortran.py [hunk 12]
```diff
@@ -1081,23 +1101,28 @@ def get_source_item(self):
         if self.isf77 and not is_f2py_directive:
             # Fortran 77 is easy..
             lines = [line[6:72]]
-            while _is_fix_cont(self.get_next_line(ignore_empty=True, ignore_comments=True)):
+            # get_next_line does not actually consume lines - they are put
+            # into the FILO buffer as well as being returned
+            while _is_fix_cont(self.get_next_line(ignore_empty=True,
+                                                  ignore_comments=True)):
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/fparser/readfortran.py]: Surprisingly, no, this is correct. I have elaborated further in the comment but basically this piece of code just deals with constructing a single line from multiple continued lines. Once that's done, parsing continues by pulling items out of the FIFO and it is then that we get any comments.

File: src/fparser/Fortran2003.py (hunk 2 above)
Comment (arporter) [reply src/fparser/Fortran2003.py]: I'm unsure :-) Anyway, I've made it explicit.

File: src/fparser/Fortran2003.py (hunk 3 above)
Comment (arporter) [reply src/fparser/Fortran2003.py]: I'd adopted the principle of just documenting what I understood and leaving anything I didn't (on the basis that there was no interface doc at all before).

File: src/fparser/Fortran2003.py (hunk 4 above)
Comment (arporter) [reply src/fparser/Fortran2003.py]: I've added to an existing test and comments after an 'end program' at least are picked up. However, as you comment below, we really need tests for this for all types of block.

**Comment by arporter** [comment]
I've now updated the documentation and added the beginnings of a developers' guide. I probably need to add more to this before giving it back for review.

**Comment by arporter** [comment]
Ready for 2nd review...

**Pull request merged**